common/src/chain/upgrades/netupgrade.rs

```diff
@@ -15,43 +15,35 @@
 
 use std::ops::Range;
 
-use crate::primitives::BlockHeight;
+use itertools::Itertools;
+
+use crate::chain::pos::{
+    DEFAULT_BLOCK_COUNT_TO_AVERAGE, DEFAULT_MATURITY_DISTANCE, DEFAULT_TARGET_BLOCK_TIME,
+};
+use crate::chain::{pos_initial_difficulty, PoSChainConfig};
+use crate::primitives::per_thousand::PerThousand;
+use crate::Uint256;
+use crate::{
+    chain::{config::ChainType, pow::limit},
+    primitives::BlockHeight,
+};
+
+use super::{ConsensusUpgrade, NetUpgradeVersion, PoSStatus, PoWStatus, RequiredConsensus};
 
 #[derive(Debug, Clone)]
 pub struct NetUpgrades<T>(Vec<(BlockHeight, T)>);
```

Comment:
Everything seems to be OK, but the only problem left is that this generic T shouldn't be part of the upgrades.

The data to be used on the upgrades should be available in the module doing the checking. Another option is to also put it in ChainConfig, which is what bitcoin does, btw. The problem here is that there's no generic solution that can contain all possible upgrades with their parametrization. This parametrization doesn't scale. The only reason this works now is that we only have consensus upgrades. In the future, if we have more, it won't fit in here.